text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,322 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This section discusses known drawbacks of the proposal as presently designed and (where applicable) plans for mitigating them in the future.
+
+## Cannot migrate off of impl Trait
+
+In this RFC, if you use `-> impl Trait` in a trait definition, you cannot "migrate away" from that. In other words, we cannot evolve:
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+```
+
+into 
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    type IntoIter: Iterator<Item = Self::Item>;
+    fn into_iter(self) -> Self::IntoIter;
+}
+```
+
+without breaking semver compatibility. The [future possibilities](#future-possibilities) section discusses one way to resolve this, by permitting impls to elide the definition of associated types whose values can be inferred from a function return type.
+
+## Clients of the trait cannot name the resulting associated type, limiting extensibility
+
+[As @Gankra highlighted in a comment on this RFC][gankra], the traditional `IntoIterator` trait permits clients of the trait to name the resulting iterator type and apply additional bounds:
+
+[gankra]: https://github.com/rust-lang/rfcs/pull/3193#issuecomment-965505149
+
+```rust
+fn is_palindrome<Iter, T>(iterable: Iter) -> bool
+where
+    Iter: IntoIterator<Item = T>,
+    Iter::IntoIter: DoubleEndedIterator,
+    T: Eq;
+```
+
+The `NewIntoIterator` trait used as an example in this RFC, however, doesn't support this kind of usage, because there is no way for users to name the `IntoIter` type (and, as discussed in the previous section, there is no way for users to migrate to a named associated type, either!). The same problem applies to async functions in traits, which sometimes wish to be able to [add `Send` bounds to the resulting futures](https://rust-lang.github.io/async-fundamentals-initiative/evaluation/challenges/bounding_futures.html).
+
+The [future possibilities](#future-possibilities) section discusses a planned extension to support naming the type returned by an impl trait, which could work to overcome this limitation for clients.
+
+## Impls cannot add new methods to the resulting types
+
+Similarly to the previous point, [@Gankra also pointed out][gankra] that impls which employ `-> impl Trait` are not able to add methods or implement traits on the resulting types. With `IntoIterator`, one has the option of adding inherent (or trait) methods to the resulting type:
+
+```rust
+struct MyIterator {
+    ...
+}
+
+impl IntoIterator for MyType {
+    type IntoIter = MyIterator;
+    ...
+}
+
+impl MyIterator {
+    fn extra_method(&self) {
+
+    }
+}
+```
+
+Using `-> impl Trait` tends to make that not work. One way to address this would be to permit impls to specify the return types of `-> impl Trait` as concrete types.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
Why `-> impl Trait` before `type Foo = impl Trait`?

From a users perspective, it sounds like an impl associated type covers a superset of use cases compared to impl in return position.  I am not familiar with the language design or compiler implementation concerns but it sounds like an impl associated type needs to be implemented along the way to impl in return position.

Reply:
`type Foo = impl Trait` is already implemented under [RFC 2515](https://github.com/rust-lang/rust/issues/63063).